### PR TITLE
Android11 Basic Support

### PR DESCRIPTION
**Pull request by deviant-aut**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
```
Comment (JeffLIrion) [inline .github/workflows/main.yml:1]: What is this file for?

File: androidtv/basetv/basetv.py [hunk 2]
```diff
@@ -128,6 +128,9 @@ def _cmd_audio_state(self):
         if constants.CUSTOM_AUDIO_STATE in self._custom_commands:
             return self._custom_commands[constants.CUSTOM_AUDIO_STATE]
 
+        # Is this a Android 11 Android TV?
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:131]:
```suggestion
        # Is this an Android 11 device?
```

Please update the other occurrences, as well.

File: androidtv/basetv/basetv.py [hunk 3]
```diff
@@ -128,6 +128,9 @@ def _cmd_audio_state(self):
...
+        if self.DEVICE_ENUM == constants.DeviceEnum.ANDROIDTV and self.device_properties.get("sw_version", "") == 11:
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:132]:
```suggestion
        if self.DEVICE_ENUM == constants.DeviceEnum.ANDROIDTV and self.device_properties.get("sw_version", "") == "11":
```

Please update the other occurrences, as well.

File: tests/test_constants.py [hunk 4]
```diff
@@ -106,6 +124,11 @@ def test_constants(self):
             constants.CMD_HDMI_INPUT,
             r"dumpsys activity starter | grep -E -o '(ExternalTv|HDMI)InputService/HW[0-9]' -m 1 | grep -o 'HW[0-9]'",
         )
+        # CMD_HDMI_INPUT11 = (
```
Comment (JeffLIrion) [inline tests/test_constants.py:127]:
```suggestion

        # CMD_HDMI_INPUT11 = (
```

File: androidtv/constants.py [hunk 5]
```diff
@@ -75,20 +75,37 @@ class DeviceEnum(IntEnum):
 #: Get the audio state
 CMD_AUDIO_STATE = r"dumpsys audio | grep paused | grep -qv 'Buffer Queue' && echo -e '1\c' || (dumpsys audio | grep started | grep -qv 'Buffer Queue' && echo '2\c' || echo '0\c')"
 
+#: Get the audio state for an Android 11 Device
+CMD_AUDIO_STATE11 = (
+    "CURRENT_AUDIO_STATE=$(dumpsys audio | sed -r -n '/[0-9]{2}-[0-9]{2}.*player piid:.*state:.*$/h; ${x;p;}') && "
+    + r"echo $CURRENT_AUDIO_STATE | grep paused >/dev/null 2>&1 && echo -e '1\c' || { echo $CURRENT_AUDIO_STATE | grep started >/dev/null 2>&1 && echo '2\c' || echo '0\c' ; }"
```
Comment (JeffLIrion) [inline androidtv/constants.py:81]:
```
grep paused >/dev/null 2>&1
```

Why not this?

```
grep -q paused
```

File: .github/workflows/main.yml (hunk 1 above)
Comment (deviant-aut) [reply .github/workflows/main.yml]: Sorry I messed up, I reverted this

File: androidtv/constants.py (hunk 5 above)
Comment (deviant-aut) [reply androidtv/constants.py]: agree

**Comment by deviant-aut** [comment]
At the moment I am somehow stuck with the device specific tests.

File: androidtv/basetv/basetv.py [hunk 6]
```diff
@@ -128,6 +128,9 @@ def _cmd_audio_state(self):
         if constants.CUSTOM_AUDIO_STATE in self._custom_commands:
             return self._custom_commands[constants.CUSTOM_AUDIO_STATE]
 
+        # Is this an Android 11 device??
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:131]:
```suggestion
        # Is this an Android 11 device?
```

File: androidtv/basetv/basetv.py [hunk 7]
```diff
@@ -149,6 +152,9 @@ def _cmd_current_app(self):
             and "Chromecast" in self.device_properties.get("model", "")
         ):
             return constants.CMD_CURRENT_APP_GOOGLE_TV
+        # Is this an Android 11 device??
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:155]:
```suggestion
        # Is this an Android 11 device?
```

File: androidtv/basetv/basetv.py [hunk 8]
```diff
@@ -186,6 +195,10 @@ def _cmd_hdmi_input(self):
         if constants.CUSTOM_HDMI_INPUT in self._custom_commands:
             return self._custom_commands[constants.CUSTOM_HDMI_INPUT]
 
+        # Is this an Android 11 device??
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:198]:
```suggestion
        # Is this an Android 11 device?
```

File: androidtv/basetv/basetv.py [hunk 9]
```diff
@@ -171,6 +177,9 @@ def _cmd_current_app_media_session_state(self):
             and "Chromecast" in self.device_properties.get("model", "")
         ):
             return constants.CMD_CURRENT_APP_MEDIA_SESSION_STATE_GOOGLE_TV
+        # Is this an Android 11 device??
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:180]:
```suggestion
        # Is this an Android 11 device?
```

File: androidtv/constants.py [hunk 10]
```diff
@@ -75,20 +75,37 @@ class DeviceEnum(IntEnum):
 #: Get the audio state
 CMD_AUDIO_STATE = r"dumpsys audio | grep paused | grep -qv 'Buffer Queue' && echo -e '1\c' || (dumpsys audio | grep started | grep -qv 'Buffer Queue' && echo '2\c' || echo '0\c')"
 
+#: Get the audio state for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:78]:
```suggestion
#: Get the audio state for an Android 11 device
```

File: androidtv/constants.py [hunk 11]
```diff
@@ -75,20 +75,37 @@ class DeviceEnum(IntEnum):
...
+    + r"echo $CURRENT_AUDIO_STATE | grep -q paused && echo -e '1\c' || { echo $CURRENT_AUDIO_STATE | grep -q started && echo '2\c' || echo '0\c' ; }"
+)
+
 #: Determine whether the device is awake
 CMD_AWAKE = "dumpsys power | grep mWakefulness | grep -q Awake"
 
 #: Parse current application identifier from dumpsys output and assign it to ``CURRENT_APP`` variable (assumes dumpsys output is momentarily set to ``CURRENT_APP`` variable)
 CMD_PARSE_CURRENT_APP = "CURRENT_APP=${CURRENT_APP#*ActivityRecord{* * } && CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && CURRENT_APP=${CURRENT_APP%\\}*}"
 
+#: Parse current application for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:90]:
```suggestion
#: Parse current application for an Android 11 device
```

File: androidtv/constants.py [hunk 12]
```diff
@@ -75,20 +75,37 @@ class DeviceEnum(IntEnum):
...
+CMD_PARSE_CURRENT_APP11 = "CURRENT_APP=${CURRENT_APP%%/*} && CURRENT_APP=${CURRENT_APP##* }"
+
 #: Assign focused application identifier to ``CURRENT_APP`` variable
 CMD_DEFINE_CURRENT_APP_VARIABLE = (
     "CURRENT_APP=$(dumpsys window windows | grep -E 'mCurrentFocus|mFocusedApp') && " + CMD_PARSE_CURRENT_APP
 )
+#: Assign focused application identifier to ``CURRENT_APP`` variable for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:97]:
```suggestion
#: Assign focused application identifier to ``CURRENT_APP`` variable for an Android 11 device
```

File: androidtv/constants.py [hunk 13]
```diff
@@ -75,20 +75,37 @@ class DeviceEnum(IntEnum):
...
+CMD_DEFINE_CURRENT_APP_VARIABLE11 = (
+    "CURRENT_APP=$(dumpsys window windows | grep 'Window #1') && " + CMD_PARSE_CURRENT_APP11
+)
+
 
 #: Output identifier for current/focused application
 CMD_CURRENT_APP = CMD_DEFINE_CURRENT_APP_VARIABLE + " && echo $CURRENT_APP"
 
+#: Output identifier for current/focused application for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:106]:
```suggestion
#: Output identifier for current/focused application for an Android 11 device
```

File: androidtv/constants.py [hunk 14]
```diff
@@ -101,6 +118,11 @@ class DeviceEnum(IntEnum):
 CMD_HDMI_INPUT = (
     "dumpsys activity starter | grep -E -o '(ExternalTv|HDMI)InputService/HW[0-9]' -m 1 | grep -o 'HW[0-9]'"
 )
+#: Get the HDMI input for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:121]:
```suggestion

#: Get the HDMI input for an Android 11 device
```

File: androidtv/constants.py [hunk 15]
```diff
@@ -117,6 +139,11 @@ class DeviceEnum(IntEnum):
     CMD_DEFINE_CURRENT_APP_VARIABLE.replace("{", "{{").replace("}", "}}") + " && " + CMD_LAUNCH_APP_CONDITION
 )
 
+#: Launch an app if it is not already the current app on an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:142]:
```suggestion
#: Launch an app if it is not already the current app on an Android 11 device
```

File: androidtv/constants.py [hunk 16]
```diff
@@ -133,6 +160,9 @@ class DeviceEnum(IntEnum):
 #: Determine the current app and get the state from ``dumpsys media_session``
 CMD_CURRENT_APP_MEDIA_SESSION_STATE = CMD_CURRENT_APP + " && " + CMD_MEDIA_SESSION_STATE
 
+#: Determine the current app and get the state from ``dumpsys media_session`` for an Android 11 Device
```
Comment (JeffLIrion) [inline androidtv/constants.py:163]:
```suggestion
#: Determine the current app and get the state from ``dumpsys media_session`` for an Android 11 device
```

File: androidtv/basetv/basetv.py [hunk 17]
```diff
@@ -149,6 +152,9 @@ def _cmd_current_app(self):
             and "Chromecast" in self.device_properties.get("model", "")
         ):
             return constants.CMD_CURRENT_APP_GOOGLE_TV
+        # Is this an Android 11 device?
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:155]:
```suggestion

        # Is this an Android 11 device?
```

File: androidtv/basetv/basetv.py [hunk 18]
```diff
@@ -171,6 +177,9 @@ def _cmd_current_app_media_session_state(self):
             and "Chromecast" in self.device_properties.get("model", "")
         ):
             return constants.CMD_CURRENT_APP_MEDIA_SESSION_STATE_GOOGLE_TV
+        # Is this an Android 11 device?
```
Comment (JeffLIrion) [inline androidtv/basetv/basetv.py:180]:
```suggestion

        # Is this an Android 11 device?
```

**Comment by JeffLIrion** [comment]
Thanks!

**Pull request merged**